cirq/api/google/v2/program.proto

```diff
@@ -0,0 +1,142 @@
+message Moment {
+  // All of the operations in the moment. Each operation must act on different
+  // qubits.
+  repeated Operation operations = 1;
+}
+
+// The quantum circuit, specified as a series of operations at specific
+// start times.
+message Schedule {
+  // A list of all the operations and their absolute start times.
+  repeated ScheduledOperation scheduled_operations = 3;
+}
+
+// An operation occurring at a specific start time.
+message ScheduledOperation {
+  // Which operation is to be scheduled.
+  Operation operation = 1;
+
+  // The start time of the operation, with zero representing the absolute
+  // start of the circuit.
+  //
+  // This must be consistent with the moment structure and must be positive.
+  int64 start_time_picos = 2;
+}
+
+// The language in which the program is expressed.
+message Language {
+  // The name of the gate set being used.
+  string gate_set = 1;
+
+  // The language supported by ArgFunctions.
+  string arg_function_language = 2;
+}
+
+// An operation acts on a set of qubits.
+message Operation {
+  // Which gate this operation corresponds to.
+  Gate gate = 1;
+
+  // Arguments needed to fully specify the gate.
+  repeated Arg args = 2;
+
+  // Which qubits the operation acts on.
+  repeated Qubit qubits = 3;
+
+  // If the operation produces results, these are referenced by these keys.
+  repeated string keys = 4;
+}
+
+// The instruction identifying the action taken on the quantum computer.
+message Gate {
+  // Name for the Gate, such as 'x' or 'exp_x' or 'x_pow'.
+  string id = 1;
+}
+
+// An identifier for a qubit.
+message Qubit {
+  // Identifies will depend on the device being scheduled upon.
+  // These depend on the device being scheduled upon.
+  //
+  // Typically ids for qubits on a line are simple string versions of integers,
+  // while for qubits on a square grid these are integers separated by a
+  // underscore, i.e. '0_1', '1_2', etc.
+  string id = 2;
+}
+
+// Arguments needed to specify a gate.
+message Arg {
+  // Arguments are either a number, a symbol, or an argument function
+  // (which recursively depends on Arg).
+  //
+  // Number is used to specify an argument that does not vary depending on
+  // RunContext.
+  //
+  // Symbol is used when an argument will be resolved (supplied a value)
+  // by a Run Context.
+  //
+  // Functions are used to define a simple s-expression tree describing
+  // how to combine numbers and symbols mathematically.
+  oneof arg {
+    float number = 1;
+    string symbol = 2;
+    ArgFunction func = 3;
+  }
+}
+
+// A function of arguments. This is an s-expression tree representing
+// mathematically the function being evaluated.
+//
+// What language is supported is specified by the arg_function_language
+// in the language message.
+message ArgFunction {
+  // The name of the function. I.e. if the function is the sum of two symbols,
+  // this could be '+', and the args would be two string symbol values.
+  string type = 1;
```

Comment:
In general, making everything a string gives you additional flexibility, which is a double-edged sword.  You can easily augment, but you also can't tell what is a valid value or what is a supported operation.

Reply:
Yeah I agree.  I think long term this becomes an enum and we are more structured here.  But I believe we will be needing to move fast for a bit and string gives the extra benefit of service supporting before exposing gate set in cirq.